experimental/incremental/executor.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package incremental
+
+import (
+	"context"
+	"runtime"
+	"slices"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Executor is a caching executor for incremental queries.
+//
+// See [New], [Run], and [Invalidate].
+type Executor struct {
+	dirty sync.RWMutex
+	tasks sync.Map // [string, *task]
+
+	sema *semaphore.Weighted
+}
+
+// New constructs a new executor with the given maximum parallelism.
+//
+// Setting parallelism to zero or negative will default to GOMAXPROCS.
+func New(parallelism int) *Executor {
+	if parallelism <= 0 {
+		parallelism = runtime.GOMAXPROCS(0)
+	}
+
+	return &Executor{
+		sema: semaphore.NewWeighted(int64(parallelism)),
+	}
+}
+
+// Queries returns a snapshot of the URLs of which queries are present (and
+// memoized) in an Executor.
+//
+// The returned slice is sorted.
+func (e *Executor) Queries() (urls []string) {
+	e.tasks.Range(func(url, t any) bool {
+		task := t.(*task) //nolint:errcheck // All values in this map are tasks.
+		result := task.result.Load()
+		if result == nil || !closed(result.done) {
+			return true
+		}
+		urls = append(urls, url.(string))
+		return true
+	})
+
+	slices.Sort(urls)
+	return
+}
+
+// Run executes a set of queries on this executor in parallel.
+//
+// This function only returns an error if ctx is cancelled during execution,
+// in which case it returns nil and the cause for cancellation.
+//
+// Errors that occur during each query are contained within the returned results.
+// Unlike [Resolve], these contain the *transitive* errors for each query!
+func Run[T any](ctx context.Context, e *Executor, queries ...Query[T]) (results []Result[T], cancelCause error) {
+	e.dirty.RLock()
+	defer e.dirty.RUnlock()
+
+	root := Task{
+		ctx:    ctx,
+		exec:   e,
+		result: &result{done: make(chan struct{})},
+	}
+
+	go func() {
+		results = Resolve(root, queries...)
+		close(root.result.done)
```

Comment:
I think this needs to be in a defer. Otherwise, if a query fails, the goroutine will exit, but since the context is not yet cancelled and this channel not closed, the select below would hang.

(Having said that, I've left other remarks about how I think the use of `runtime.Goexit` is dangerous and we should probably revisit how failures are communicated. But this probably needs to be a deferred function no matter how those changes play out.)

Reply:
You're right regardless, if a panic comes crashing through for whatever reason, it will get swallowed.